src/core/Tools.cpp

```diff
@@ -379,12 +380,12 @@ namespace Tools
         return result;
     }
 
-    QString
-    substituteBackupFilePathPattern(QString pattern, const QString& databasePath, QDateTime date, int maxSubstitutions)
```

Comment:
Should we not maybe keep date and maxSubstitution parameters? Date generalizes the function which could be useful if we later want to use it in a different context. maxSubstitution is actually required for correctness if we add another regex replacement. As soon as we do that we could have loop by substituting A with B and vice versa (see the tests for a sketch). We could just hardcode the value for now though, or at least put a warning comment there.

Reply:
I made it impossible to recurse regrx by not doing a global match all at once. It matches the string after each replacement looking for the patterns, which will overwrite recursive ones with the substitution instead. 

It doesn't make any sense to have any other time besides now. A time in the past would either be a calculation from now (now - 1 day) or a an absolute date which wouldn't be a substitution.

Reply:
Sorry, i don't get how match() vs globalMatch() fixes a recursion error. We were replacing matches one-at-a-time (left-to-right) with globalMatch() in the previous code version, reevaluating after every substitution. Could you perhaps explain? Not saying that using match() is not more appropriate here, just curious what i am missing :sweat_smile:.

I was trying to address the following problem with the maxSubstitutions parameter: assume that we modify the function to add other patterns/use it in a different context (it's in the Tool namespace now, so i guess that will probably happen at some point). Maybe we add a placeholder that substitutes some DB Attribute, like the Database description or name. Then, setting the description to {DB_ATTR:DESC} and supplying regex "{DB_ATTR:DESC}" would cause an infinite loop because "{DB_ATTR:DESC}" is replaced with itself.

Again, this cannot happen right now. If we still want to address this issue and make the function more robust for future modifications we could either i) add the maxSubstitutions parameter, or ii) advance the starting point of match() in the new string based on the current index and length of the consumed substring + replacement to not consider already replaced placeholders.

Reply:
Globalmatch finds all matches at once, including recursive ones, and then you iterate that match list (QRegularExpressionMatchIterator). You were not finding them one at a time, that is what match does by returning a QRegularExpressionMatch instead.

Reply:
> I made it impossible to recurse regex by not doing a global match all at once. It matches the string after each replacement looking for the patterns from left to right, which will overwrite recursive ones with the substitution instead.

I get that global match finds all matches at once and that it's unnessesary in this case and we should use match() instead. But even when using globalMatch() it is impossible to recurse regex, no? I don't see how substituting match() for globalMatch() modifies the behavior of the function given that i was only using the first match found previously.

Reply:
This is true based on your previous implementation since you re-performed the global match after each replacement.